### ISSUE TITLE
nginx fork stubbed

### ISSUE THREAD

**Issue by leebaok** [cloudius-systems/osv-apps#57]
The nginx app runs failed with the error: **'fork() stubbed, nginx:[emerg] fork() failed'**

There is a patch to replace fork in src/os/unix/ngx_process.c while another fork is still in **src/os/unix/ngx_daemon.c**.  So, maybe the **fork in ngx_daemon.c** need to be replaced too.

**Comment by nyh** [comment]
This is a wild guess (I didn't do the nginx work myself), but maybe nginx is going for "daemon mode" where the first thing it does is to fork(), and this simply needs to be avoided? Maybe with a "daemon off;" directive in the configuration file?

**Comment by leebaok** [comment]
Thanks. The fail of fork is caused by daemon mode.

But in the daemon off mode, there is another issue:
`
libao@ubuntu-25:~/capstan/osv-apps/nginx$ sudo ../../capstan run -v -f 10000:80
Created instance: nginx
OSv v0.24
eth0: 192.168.122.15
/nginx.so: failed looking up symbol sigsuspend

[backtrace]
0x00000000003447cd <elf::object::symbol(unsigned int)+205>
0x0000000000346148 <elf::object::resolve_pltgot(unsigned int)+136>
0x0000000000346325 <elf_resolve_pltgot+69>
0x0000000000391e64 <???+3743332>
0xffff8000057f108f <???+92213391>
`

Does signal work in OSv ?

**Comment by nyh** [comment]
The latest OSv *does* have a sigsuspend symbol - it's not fully implemented, but you should not have gotten the error above. Is it possible you are using an older version of OSv? Can you please try checking out the latest git version?

**Comment by leebaok** [comment]
Hi. The output above says the version is 0.24. Is it the old one or the new one ?
(I checkout the code last week from the master branch)

**Comment by nyh** [comment]
0.24 is an antique version. It's the last official version we released, but you shouldn't use it. Please check out the master branch, not a version branch.

**Comment by leebaok** [comment]
I use the newest code from master branch. But it still fails : 

```
libao@ubuntu-25:~/osv$ sudo ./scripts/run.py -V --novnc
OSv v0.24-432-g4a71643
4 CPUs detected
Firmware vendor: Bochs
bsd: initializing - done
VFS: mounting ramfs at /
VFS: mounting devfs at /dev
net: initializing - done
vga: Add VGA device instance
eth0: ethernet address: 52:54:00:12:34:56
virtio-blk: Add blk device instances 0 as vblk0, devsize=10737418240
random: virtio-rng registered as a source.
random: <Software, Yarrow> initialized
VFS: unmounting /dev
VFS: mounting zfs at /zfs
zfs: mounting osv/zfs from device /dev/vblk0.1
random: device unblocked.
VFS: mounting devfs at /dev
VFS: mounting procfs at /proc
program zpool.so returned 1
BSD shrinker: event handler list found: 0xffffa000017be480
	BSD shrinker found: 1
BSD shrinker: unlocked, running
[I/42 dhcp]: Broadcasting DHCPDISCOVER message with xid: [27037311]
[I/42 dhcp]: Waiting for IP...
[I/244 dhcp]: Received DHCPOFFER message from DHCP server: 192.168.122.1 regarding offerred IP address: 192.168.122.15
[I/244 dhcp]: Broadcasting DHCPREQUEST message with xid: [27037311] to SELECT offered IP: 192.168.122.15
[I/244 dhcp]: Received DHCPACK message from DHCP server: 192.168.122.1 regarding offerred IP address: 192.168.122.15
[I/244 dhcp]: Server acknowledged IP 192.168.122.15 for interface eth0 with time to lease in seconds: 86400
eth0: 192.168.122.15
[I/244 dhcp]: Configuring eth0: ip 192.168.122.15 subnet mask 255.255.255.0 gateway 192.168.122.1 MTU 1500
could not load libcrypt.so.1
could not load libz.so.1
WARNING: af_local::ioctl(FIOASYNC) stubbed
WARNING: fcntl(F_SETOWN) stubbed
initgroups() stubbed
QEMU: Terminated
```
It is a little hard to run a simple nginx with OSv. 

Thanks.

**Comment by nyh** [comment]
I can't reproduce this problem. I tried the same, and it seems to work: after the "initgroups() stubbed" message (which is benign), the VM just waits forever - serving requests - as expected. It does not print the "QEMU: Terminated" message, nor terminate.  Only when I deliberately kill it with control-A x, qemu prints the "QEMU: Terminated" message and exits...

**Comment by leebaok** [comment]
Thanks. 
"QEMU: Terminated" is after my clicking control-A x. 
So "initgroups() stubbed" is the normal output ?

But I cannot connect the network of OSv VM. Can you tell me how to configure networking of OSv ? Or just use the default settings of the command "sudo ./scripts/run.py -V --novnc" ?

Thanks.

**Comment by nyh** [comment]
Yes, that "initgroups() stubbed" message is just a benign message (we didn't implement this function, but everything works correctly), and afterwards the server just silently waits for connections.

If you're using scripts/run.py to run the VM, you have two options how to set up network:

1. The simplest option, useful for testing, is to ask QEmu to forward some port on your host (e.g., port 8000) to port 80 inside the guest (which is served by nginx). This can be done with the command: "scripts/run.py --forward tcp:8000::80". With this, on your host, you can make a request to port 80 (e.g., curl http://localhost:8000/) and see the guest nginx answer!

2. Use bridged networking - have the guest connected to a bridge device so packets arriving to the bridge are sent to the guest, and packets going out of the guest also get sent to the bridge. You do this with the "-n" option (and optionally also the "-v" option for the faster vhost), but you may need to read up on the Internet on how to properly set up such a bridge (you'll also need a DHCP server on this bridge) - or have something like libvirt do it for you.

I'm closing this issue now, as this is a tracker for bugs, not questions and discussions. I'll be happy to continue this discussion, on the OSv mailing list: https://groups.google.com/forum/#!forum/osv-dev
